### PR TITLE
JavaScript: Reduce extractor memory requirement.

### PR DESCRIPTION
**Pull request by xiemaisi**
During multi-threaded extraction, all extractor threads share the same heap, which sometimes causes out-of-memory errors on projects that can be extracted successfully in single-threaded mode, notably `node` and `test262` which are part of our standard benchmark suite, but also seven other projects on LGTM.com (see [internal issue](https://jira.semmle.com/browse/ODASA-7720) for details).

This PR makes the extractor less memory-hungry, allowing us to extract `node` and `test262` as well as four of the projects in the above-mentioned issue with two threads.

The main improvement is in https://github.com/Semmle/ql/commit/6013b918fc80311e4ef407e6bc365388bc31b275: we now extract lexical information before AST information, so we can discard token and comment information early and don't need to hold on to it for the rest of the extraction.

The only obstacle to this is that we use token information to compute lines-of-code information for functions, but we can just as well do that in QL, so that's what https://github.com/Semmle/ql/commit/a480c6ecaac20776525ac40fe038d69630d01832 implements.

While testing that the QL-based implementation gives the same results as the extractor-based one, I found an off-by-one bug in an extensional predicate used by the former, which is fixed by https://github.com/Semmle/ql/commit/d6c3ae2fb49f1a62245d394a5eae85da457265c9.

Finally, https://github.com/Semmle/ql/commit/5eba486d34e82e61f3b0f22a726972899b59cd1e implements an unrelated memory optimisation in the CFG extractor, where we can clear some caches after we're done with each function.

Commit-by-commit review is essential; you'll want to skip the "Update trap tests" commits, which do just what the name suggests but aren't otherwise interesting.

### CODE REVIEW

**Comment by xiemaisi** [comment]
@asger-semmle, the last commit temporarily removes a test that doesn't pass for me on `master`, see internal JIRA issue for details.

File: javascript/extractor/src/com/semmle/js/extractor/CFGExtractor.java [hunk 1]
```diff
@@ -774,13 +774,13 @@ public Finally(SourceLocation loc, BlockStatement body) {
 		}
 
 		// associate statements with their (direct or indirect) labels
-		private final Map<Statement, Set<String>> loopLabels = new LinkedHashMap<Statement, Set<String>>();
+		private Map<Statement, Set<String>> loopLabels = new LinkedHashMap<Statement, Set<String>>();
```
Comment (esben-semmle) [inline javascript/extractor/src/com/semmle/js/extractor/CFGExtractor.java]: Perhaps document that these three caches are replaced and restored at each function.

**Comment by xiemaisi** [comment]
> Is it correct that a line can count as both a code line and a comment line?

Correct.

> Which tokens can span multiple lines?

Template literals.

File: javascript/extractor/src/com/semmle/js/extractor/CFGExtractor.java (hunk 1 above)
Comment (xiemaisi) [reply javascript/extractor/src/com/semmle/js/extractor/CFGExtractor.java]: Good plan.

**Pull request merged**